src/query/sql/src/planner/optimizer/optimizers/operator/filter/infer_filter.rs

```diff
@@ -188,6 +199,28 @@ impl<'a> InferFilterOptimizer<'a> {
             Some(index) => self.expr_equal_to[*index].push(left.clone()),
             None => self.add_expr(right, vec![], vec![left.clone()]),
         };
+
+        true
+    }
+
+    // equal expr must have the same type, otherwise the function may fail on execution.
+    fn check_equal_expr_type(left_ty: &DataType, right_ty: &DataType) -> bool {
```

Comment:
Do we need(I am not sure other places has) some infer types compare?
Like Number Type cmp  Decimal Type?

Reply:
Yes, you're right that this approach might miss some potential optimizations. But the functions supporting `Number` types might not be directly applicable to `Decimal` types. Fully supporting these kinds of cross-type inferences would require a more complex logic to ensure type compatibility. I haven't yet devised a clean and robust solution for handling these scenarios. We can consider exploring more advanced type inference strategies as a future enhancement.

Reply:
This PR is OK, let's tracking this in new issue.